morl_baselines/multi_policy/pareto_q_learning/pql_stochastic.py

```diff
@@ -0,0 +1,312 @@
+        else:
+            values = []
+            for action in range(self.num_actions):
+                q_set = self.calc_q_set(state, action)
+                q_set = get_non_dominated(q_set)
+                value = self.perf_indic(q_set)
+                values.append(value)
+            action = np.random.choice(np.argwhere(values == np.max(values)).flatten())
+            return action
+
+    def update(self, state, action, next_state, r):
+        """Perform an update of the learner.
+
+        Args:
+            state (int): The previous state.
+            action (int): The action that was taken.
+            next_state (int): The next state.
+            r (ndarray): The reward vector for this transition.
+
+        Returns:
+            None
+        """
+        self.transitions[state, action, next_state] += 1
+        q_sets = []
+        for a in range(self.num_actions):
+            q_sets.append(self.calc_q_set(next_state, a))
+        self.non_dominated[state][action][next_state] = get_best(set().union(*q_sets), max_points=self.novec)
+        self.avg_r[state, action, next_state] += (r - self.avg_r[state, action, next_state]) / self.transitions[
+            state, action, next_state
+        ]
+
+    def construct_pcs(self):
+        """Construct the Pareto Coverage Set.
+
+        Returns:
+            List[List[Set]]: A set of undominated rewards per state-action pair.
+        """
+        pcs = [[{tuple(np.zeros(self.num_objectives))} for _ in range(self.num_actions)] for _ in range(self.num_states)]
+        for state in range(self.num_states):
+            for action in range(self.num_actions):
+                pcs[state][action] = get_non_dominated(self.calc_q_set(state, action))
+        return pcs
+
+    def construct_pf(self):
+        """Construct the Pareto front.
+
+        Returns:
+            Set: A set of rewards which are on the Pareto front.
+        """
+        q_sets = []
+        for action in range(self.num_actions):
+            q_set = self.calc_q_set(0, action)
+            q_sets.append(q_set)
+        pareto_front = set().union(*q_sets)
+        pareto_front = get_non_dominated(pareto_front)
+        return pareto_front
+
+    def get_config(self):
+        """Get the config parameters for this agent.
+
+        Returns:
+            Dict: A dictionary of configuration parameters.
+        """
+        return {
+            "env_id": self.env.unwrapped.spec.id,
+            "init_epsilon": self.init_epsilon,
+            "epsilon_decay": self.epsilon_decay,
+            "decay_every": self.decay_every,
+            "min_epsilon": self.min_epsilon,
+            "gamma:": self.gamma,
+            "decimals": self.decimals,
+            "novec": self.novec,
+        }
+
+    def flatten_observation(self, obs):
+        return int(np.ravel_multi_index(obs, (11, 11)))
+
+    def train(
+        self,
+        iterations: int = 3000,
+        max_timesteps: int = 1000,
```

Comment:
Max. timesteps per episode is controlled in the environment itself. If timestep >= max_timesteps, then the environment returns truncated=True